core/src/main/java/com/spotify/ffwd/output/CoreOutputManager.java

```diff
@@ -220,6 +292,22 @@ private boolean rateLimitAllowed(int permits, String key) {
         }
     }
 
+    private boolean cardinalityLimitAllowed(long currentCardinality) {
+        return cardinalityLimit == null || cardinalityLimit >= currentCardinality;
+    }
+
+    /**
+    * To reset cardinality this will swap HLL++ if it was tripped after configured period of ms
+    */
+    private void swapHyperLogLogPlus() {
+        if (System.currentTimeMillis() - hyperLogSwapTS.get() > hyperLogLogPlusSwapPeriod
+            && hyperLogSwapLock.compareAndExchange(false, true)) {
+          hyperLog = new HyperLogLogPlus(14, 25);
```

Comment:
i might be wrong but I think you need a `synchronized` bit here so all threads will see the new reference. i need to come back to this part.

Reply:
I've updated this to use AtomicReference instead of an object itself.